modules/optimization/src/executioners/OptimizeSolve.C

```diff
@@ -287,6 +295,8 @@ OptimizeSolve::setTaoSolutionStatus(double f, int its, double gnorm, double cnor
       steady->setIterationNumberOutput((unsigned int)its);
   }
 
+  if (_time_step_as_iteration)
+    _problem.timeStep() = its;
```

Comment:
This would allow for outputting at each iteration. I couldn't find a better way to set this.

Reply:
Hmm... I don't think we want to do this. This set would be frustrating to track down in the future

Reply:
Any idea how we can do this. This would be a really useful debugging option because it would keep track of the current iteration, when the sub app doesn't solve. For exodus files when executing on `LINEAR` it increments the time by a very small number. But I can't find where it does that.

Reply:
there were talks of having a naming convention for outputs like
```
...._t<timestep number>_nl<nl ite number_l<linear it number>
```
we could do the same here for forward
@lindsayad was involved iirc

Reply:
What Max has also allows us to output exodus files on the main app and they will be incremented by the timestep so we can then view them in paraview.  I think we need the time to change because exodus does have a "time_step" but it doesn't have an "iteration" for saving different model states.   This PR is provides a really important debugging capability for optimization problems.

Reply:
the time is already changing, with the tiny increments (0.0...01 depending on the nesting)

Reply:
So update nonlinear iterations instead of the timestep and good to go?

Reply:
https://github.com/idaholab/moose/issues/25416 is the issue I believe. I think I'm the new Derek over-promising what I can do ... I said I would work on this many months ago but have not done so and likely will not be able to this fiscal year so if someone else wants to take that over or circumvent it if there's another good solution then I'm not opposed

Reply:
I'll sign off on abusing nonlinear iteration counts. Someone debugging nonlinear iterations within an optimization solve will just have to get wrekt or change the code

Reply:
So a few things. 
1) These things go into the main app for an optimization problem which never solves the physics. It would be really hard to solve the physics and opt problem together with the way it is setup.   
2) So the only Optimize executioner is a steady executioner so if I can change the _t_step which is what I have currently all the outputting works really nice. 

Looking at Output.C, the way we get the time is based off the _t_step. 
``` 
Real
Output::time()
{
  if (_transient)
    return _time;
  else
    return _t_step;
}
``` 
If we did it any other way we would have to add a lot more hacky code.
For example in `JsonOutput.C` to output the nonlinear iteration as the tilmestep I would need to add a `contains(OptimizationAppTypes::EXEC_FORWARD)` if statement to make it work for Json fils, then something for csv files, the exodus files, etc.
```
   if (_execute_enum.contains(EXEC_LINEAR) && !_on_nonlinear_residual)
      current_node["linear_iteration"] = _linear_iter;
    if (_execute_enum.contains(EXEC_NONLINEAR))
      current_node["nonlinear_iteration"] = _nonlinear_iter;
```

Reply:
Is anything done with this executioner using transient at the moment?

Reply:
Even for transient optimizations we use this executioner for the main app. The "transient" part lives on the sub apps only.

Reply:
Part of me wants to say just kick this down the road to get it in and not let it work with transient?

Reply:
Alright seems fair to me. I dont think we have the resource to look at this

Reply:
Yeah, I'm on board with it as long as @maxnezdyur isn't interrupted by it and doesn't see it as a near-future need.

Reply:
Added an error check so it can't work with transient executioners. I agree getting this in sooner will provide a lot of value.